state: track substate ComputedVar that depends on parent Var

masenf opened the pull request:
Follow on from #840 and #917, that introduces substate dependence determination at `__init__` time so that each state tracks which variables its substates depend on. When marking a state dirty, if any dirty vars are dependencies of substate ComputedVar, then pass the dirty var down to the substate and also mark it dirty.

Fix #839 

### All Submissions:

- [x] Have you followed the guidelines stated in [CONTRIBUTING.md](https://github.com/pynecone-io/pynecone/blob/main/CONTRIBUTING.md) file?
- [x] Have you checked to ensure there aren't any other open [Pull Requests](https://github.com/pynecone-io/pynecone/pulls ) for the desired changed?

### Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

### New Feature Submission:

- [x] Does your submission pass the tests? 
- [x] Have you linted your code locally prior to submission?

### Changes To Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully ran tests with your changes locally?

Edit 2023-04-23: rebased after 840 merged, diff should be clean now and ready for review!
Edit 2023-05-03: rebased after #917 merged, diff is now clean and ready for review!

Thanks for taking my previous patch! 🙏

picklelo commented:
@masenf I put up a PR to speed up the dependency management: https://github.com/pynecone-io/pynecone/pull/864/files

Once that's merged, can you rebase this logic into the new format? Appreciate all your contributions!!

masenf commented:
Now that computed vars dependencies are tracked all the way to the base state, there's no need to track explicit substate var dependencies...can just push all of the `dirty_vars` down into each substate and let the substate determine which vars need to be recomputed.

This might still result in calculating some _parent_ computed var multiple times, especially if lots of other variables depend on it. Would need to poke around and see if computed var values get cached already or not. The problem is solvable because the nature of substate tree means parent computed vars will always be recalculated before substate is called.

masenf commented:
@picklelo i think we can get away with removing `dirty_substate` tracking altogether!

picklelo commented on pynecone/state.py:
```diff
@@ -706,32 +703,24 @@ def get_delta(self) -> Delta:
-        for substate in self.dirty_substates:
-            delta.update(substates[substate].get_delta())
+        # Recursively find all substate deltas.
+        for substate in self.substates.values():
```
I think we may want to keep the `dirty_substates` prop. This will work, but as the substate tree grows very big, this may end up having some performance impact as we will call `get_delta()` for every substate in the tree.

picklelo reviewed (commented):
Nice! I left one comment - I think the `dirty_substates` prop may be useful for performance reasons.

I have this PR #872  that temporarily disables the computed var dependency checks until we have the static checks from #870 so we don't have any breaking changes.

We can either merge this PR in before #870 with the checks disabled for now, or wait until we have that, either way should be okay.

masenf commented:
The issue that I was having with dirty_substates was how to know when a grandchild that depended on a parent var needs to update?

I was passing all of the dirty vars into each substate (to pass into their substates) because the child state doesn't know what the grandchild computed var dependencies are without checking (and it was getting complicated).

I ended up simplifying by removing dirty_substates because my approach basically put all substates in the dirty set, so it made more sense up exhaustively iterate, and have each state only responsible for its own computed var deps.

I'll try to rework this with more perf consideration in mind, but I don't want to do it by making the code excessively complicated, unless a clear perf benefit is noted.

picklelo commented:
@masenf I just tested it out on our [pcweb](https://github.com/pynecone-io/pcweb) repo since it has a huge state tree, and the process time for state updates went from around `0.0010` to `0.0026` so I think there is some perf impacts.

The `mark_dirty` method recursively calls `self.parent_state.mark_dirty()` so the grandchild state should mark all the states up to the parent dirty. Can we add the grandchild computed var in the parent var dependencies - and when the parent var updates we can mark the grandchild as dirty?

masenf commented:
I'm working on a "deep state" test that has 1000+ state nodes for doing deterministic perf testing.

I'm wondering if the extra time on pcweb is due to inefficiency in the algorithm, or the fact that substate delta was just not recomputed at all before this change.

picklelo commented:
Yeah I don't think substate computed vars that only depended on parent vars were computed before. If the performance impact is too large, maybe this limitation is okay since it may be a rare use case. But if there's a clean solution would be good to include.

masenf commented:
@picklelo the latest update reduces the process time by an order of magnitude over my previous patch. still not ready to go in (docstrings, tests, etc), but I wanted to post my progress thus far. The main wins comes from calculating the explicit var -> substate dependencies _after_ tracking during `__init__` and pushing those relationships into each parent state. Then when `mark_dirty` is called, it only has to mark substates dirty that actually depend on the changed vars, which makes a significant difference for variables that only have 1 deep dependency.

One of the things that's going to really hurt us in real world perf is recalculating the ComputedVar everytime it's accessed by some dependency. For high-level, expensive computed vars, this could make things _really_ slow if a lot of substates depend on it.

masenf committed:
state: propagate new variables to substates

fixes pynecone-io#893

masenf committed:
state: track substate var dependencies

update deep child states when dependent parent Var is modified

masenf committed:
state: move inherited_var calculation outside of loop

There's no reason to calculate the set of inherited_vars +
backend_inherited_vars for each iteration of the loop.

Also, use a py37 compatible union ('|' not supported)

masenf commented:
I'm pretty happy with how this one turned out. With this PR (including #916 and #917), there appears to be no regression processing state updates in the `pcweb` project!

Give it a shot and let me know what you think.

masenf committed:
state: assign substate_var_dependencies early in __init__

ensure that each State instance gets its own dependency tracking variables.

Assign these _early_ in __init__, so that recursively initializing child states
can access the substate_var_dependencies of the parent_state.

masenf committed:
state: move substate get_delta calculation to tail

this was moved in #917 for no significant reason.

now that caching is implemented for ComputedVar, there's no reason to start
from the "bottom" of the state tree

masenf committed:
state: stop propagating substate dependency when parent_state does not know var

fixup: stop passing dependency up into states where they wont be used; wasn't
really hurting anything other than wasting space

masenf committed:
Merge branch 'main' into substate-var-dependencies

picklelo commented on pynecone/state.py:
```diff
@@ -377,6 +397,10 @@ def add_var(cls, name: str, type_: Any, default_value: Any = None):
         cls.vars.update({name: var})
 
+        # let substates know about the new variable
+        for substate_class in cls.__subclasses__():
```
nit: we can use `cls.get_substates()` here

picklelo commented on pynecone/state.py:
```diff
@@ -724,14 +748,27 @@ def get_delta(self) -> Delta:
+            self.parent_state is not None
+            and state_name not in self.parent_state.dirty_substates
+        ):
             self.parent_state.dirty_substates.add(self.get_name())
```
We can just use `state_name` here

picklelo reviewed (approved):
nice! Just some nits we can clean up in a follow up
